### PR TITLE
Issue #4334 - Improve testing of ErrorHandler behavior

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java [hunk 1]
```diff
@@ -302,29 +385,150 @@ public void testBadMessage() throws Exception
...
+    }
+
+    private String assertContent(HttpTester.Response response)
+    {
+        String contentType = response.get(HttpHeader.CONTENT_TYPE);
+        String content = response.getContent();
+
+        if (contentType.contains("text/html"))
+        {
+            assertThat(content, not(containsString("<script>")));
```
Comment (jonare) [inline jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java:467]: The usefulness of such an attack isn't as obvious as script tags in HTML, but we should assure that (potentially malformed) JSON or XML messages in exceptions are properly escaped as well. Eg. new Exception("\"}{\"I am injected JSON\":\"evil\"}{\"void\":\"") or new Exception("</something><injectedxml>evil</injectedxml><something>"), or whatever could break out of the JSON/XML structure and create or modify elements.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 2]
```diff
@@ -481,24 +482,24 @@ private void writeErrorJson(HttpServletRequest request, PrintWriter writer, int
...
-
-
     }
 
     protected void writeErrorPageStacks(HttpServletRequest request, Writer writer)
         throws IOException
     {
         Throwable th = (Throwable)request.getAttribute(RequestDispatcher.ERROR_EXCEPTION);
-        if (_showStacks && th != null)
+        if (th != null)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java:493]: why has the test on `_showStacks` been removed?

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by gregw — APPROVED** [review]

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java (hunk 2 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java:493]: This usage is protected above.

File: jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java (hunk 1 above)
Comment (joakime) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java:467]: That's been addressed.
The ErrorHandler only produces `text/html`, `text/json`, or `text/plain` output.

For `text/json` all content from exceptions (the stacktrace, messages, causes, and supressed) are filtered via [`StringUtil.sanitizeXmlString()`](https://github.com/eclipse/jetty.project/blob/ba1fe2826d49c52f82b49bb32737a361519c0479/jetty-util/src/main/java/org/eclipse/jetty/util/StringUtil.java#L1109-L1171)

For `text/json` all values are filtered via the same `StringUtil.sanitizeXmlString()`.

For `text/plain` they are left alone.

Comment (joakime) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java:467]: I'll add another test for the JSON example you provided.

Comment (joakime) [reply jetty-server/src/test/java/org/eclipse/jetty/server/ErrorHandlerTest.java:467]: Yeah, it's fine. 
See commit 217602e97f2b9de41e042620148f910a82626cd0

The generated `text/json` looks like this ...

``` json
{
"cause1":"java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"cause0":"javax.servlet.ServletException: java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"message":"javax.servlet.ServletException: java.lang.RuntimeException: &quot;}, &quot;glossary&quot;: {\n &quot;title&quot;: &quot;example&quot;\n }\n {&quot;",
"url":"/jsonmessage/",
"status":"500"
}
```

**Pull request merged**